File: src/main/java/org/jukeboxmc/block/behavior/BlockCake.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.block.data.UpdateReason;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.GameMode;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.Difficulty;
import org.jukeboxmc.world.World;

public class BlockCake extends Block implements Waterlogable {

    public BlockCake(Identifier identifier) {
        super(identifier);
    }

    public BlockCake(Identifier identifier, NbtMap blockStates) {
        super(identifier, blockStates);
    }

    @Override
    public boolean placeBlock(Player player, World world, Vector blockPosition, Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace) {
        if (this.getSide(BlockFace.DOWN) instanceof BlockAir) {
            return false;
        }
        if (world.getBlock(placePosition) instanceof BlockWater blockWater && blockWater.getLiquidDepth() == 0) {
            world.setBlock(placePosition, Block.create(BlockType.WATER), 1, false);
        }
        world.setBlock(placePosition, this);
        return true;
    }

    @Override
    public long onUpdate(UpdateReason updateReason) {
        if (updateReason.equals(UpdateReason.NORMAL)) {
            if (this.getSide(BlockFace.DOWN) instanceof BlockAir) {
                this.location.getWorld().setBlock(this.location, Block.create(BlockType.AIR));
            }
        }
        return 0;
    }

    @Override
    public boolean interact(Player player, Vector blockPosition, Vector clickedPosition, BlockFace blockFace, Item itemInHand) {
        if ((player.getHunger() < 20 || player.getGameMode().equals(GameMode.CREATIVE) || player.getServer().getDifficulty().equals(Difficulty.PEACEFUL))) {
            int biteCounter = this.getBiteCounter();
            if (biteCounter <= 6) {
                this.setBiteCounter(biteCounter = biteCounter + 1);
            }
            if (biteCounter > 6) {
                player.getWorld().setBlock(this.location, Block.create(BlockType.AIR));
            } else {
                player.addHunger(2);
                player.setSaturation(Math.min(player.getSaturation() + 0.4f, player.getHunger()));
            }
            return true;
        }
        return false;
    }

    @Override
    public int getWaterLoggingLevel() {
        return 1;
    }

    public BlockCake setBiteCounter(int value) {
        this.setState("bite_counter", Math.min(value, 6));
        return this;
    }

    public int getBiteCounter() {
        return this.stateExists("bite_counter") ? this.getIntState("bite_counter") : 0;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockCake.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Block
import org.jukeboxmc.api.block.BlockType
import org.jukeboxmc.api.block.Cake
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.api.player.GameMode
import org.jukeboxmc.api.world.Difficulty
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockCake(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    Cake {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        if (this.getRelative(BlockFace.DOWN).getType() == BlockType.AIR) {
            return false
        }
        val block = world.getBlock(placePosition)
        if (block is BlockWater && block.getLiquidDepth() == 0) {
            world.setBlock(placePosition, block, 1, false)
        }
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun interact(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        clickedPosition: Vector,
        blockFace: BlockFace,
        itemInHand: JukeboxItem
    ): Boolean {
        if (player.isHungry() || player.getGameMode() == GameMode.CREATIVE || player.getWorld()
                .getDifficulty() == Difficulty.PEACEFUL
        ) {
            var biteCounter: Int = this.getCounter()
            if (biteCounter < 6) {
                this.setCounter(biteCounter + 1.also { biteCounter = it })
            }
            if (biteCounter >= 6) {
                player.getWorld().setBlock(this.getLocation(), AIR)
            } else {
                player.addHunger(2)
                player.addSaturation(0.4f)
            }
            return true
        }
        return false
    }

    override fun getCounter(): Int {
        return this.getIntState("bite_counter")
    }

    override fun setCounter(value: Int): BlockCake {
        return this.setState("bite_counter", value)
    }

    override fun getWaterLoggingLevel(): Int {
        return 1
    }
}